### PR TITLE
add idr(s) solver

### PR DESCRIPTION
**Pull request by tcclevenger**
This is an implementation of the preconditioned IDR(s) Krylov solver from https://dl.acm.org/citation.cfm?id=2049667. The `AdditionalData` object only takes in a parameter `s` which can be any positive integer. The reason for this solver is that it can show convergence behavior similar to GMRES for higher values of `s`, but has a short term recurrence. 

I updated lac/solver_02 test to include IDR(2) and updated the output. Let me know if I should implement a completely new test.

Here is an example using IDR(s) in step-63 (advection-diffusion, Q3, multigrid with Jacobi, one smoothing step): 
  Cycle 0:
     Number of active cells:       512 (4 levels)
     Number of degrees of freedom: 4800

  GMRES:    Solved in 92 iterations (runtime = 0.13692).
  GMRES(50): Solved in 100 iterations (runtime = 0.126711).
  BiCGStab: Solved in 73 iterations (runtime = 0.166283).
  IDR(1): Solved in 72 iterations (runtime = 0.158395).
  IDR(2): Solved in 40 iterations (runtime = 0.130106).
  IDR(3): Solved in 28 iterations (runtime = 0.123259).
  IDR(4): Solved in 22 iterations (runtime = 0.127301).
  IDR(5): Solved in 18 iterations (runtime = 0.123031).
  IDR(6): Solved in 16 iterations (runtime = 0.125664).
  IDR(7): Solved in 13 iterations (runtime = 0.121045).
  IDR(8): Solved in 12 iterations (runtime = 0.136568).
  IDR(9): Solved in 11 iterations (runtime = 0.137796).

residuals

### CODE REVIEW

File: include/deal.II/lac/solver_idr.h [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+   * A namespace for a helper class to the IDR(s) solver.
+   */
+  namespace SolverIDRImplementation
+  {
+    /**
+     * Class to hold temporary vectors whose size depends on
+     * the solver parameter s.
+     */
+    template <typename VectorType>
+    class TmpVectors
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: can you move this to the .cc file ?

File: include/deal.II/lac/solver_idr.h [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+template <class VectorType = Vector<double>>
+class SolverIDR : public SolverBase<VectorType>
+{
+public:
+  /**
+   * Structure for storing additional data needed by the solver.
+   */
+  struct AdditionalData
+  {
+    explicit AdditionalData(const unsigned int s = 2)
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: document the constructor please

File: include/deal.II/lac/solver_idr.h [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+    const unsigned int s;
+  };
+
+  /**
+   * Constructor.
+   */
+  SolverIDR(SolverControl &           cn,
+            VectorMemory<VectorType> &mem,
+            const AdditionalData &    data = AdditionalData(2));
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: remove the 2 here

File: include/deal.II/lac/solver_idr.h [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+
+  /**
+   * Constructor. Use an object of type GrowingVectorMemory as a default to
+   * allocate memory.
+   */
+  SolverIDR(SolverControl &cn, const AdditionalData &data = AdditionalData(2));
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: remove the 2 here

File: include/deal.II/lac/solver_idr.h [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+       * Get vector number @p i. Allocate it if necessary.
+       *
+       * If a vector must be allocated, @p temp is used to reinit it to the
+       * proper dimensions.
+       */
+      VectorType &
+      operator()(const unsigned int i, const VectorType &temp);
+
+      /**
+       * Return size of data vector.
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: is this the number of vectors stored?

File: include/deal.II/lac/solver_idr.h [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+      return *data[i];
+    }
+
+
+
+    template <class VectorType>
+    unsigned int
+    TmpVectors<VectorType>::size() const
+    {
+      return (data.size() > 0 ? data.size() - 1 : 0);
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: I don't understand this logic. Can you explain?

File: include/deal.II/lac/solver_idr.h [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+    return;
+
+  // Initialize sets of vectors/matrices whose size dependent on s
+  internal::SolverIDRImplementation::TmpVectors<VectorType> G(s, this->memory);
+  internal::SolverIDRImplementation::TmpVectors<VectorType> U(s, this->memory);
+  internal::SolverIDRImplementation::TmpVectors<VectorType> Q(s, this->memory);
+  FullMatrix<double>                                        M(s, s);
+
+  // Random number generator for vector entries of
+  // Q (normal distirbution, mean=0 sigma=1)
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]:
```suggestion
  // Q (normal distribution, mean=0 sigma=1)
```

File: include/deal.II/lac/solver_idr.h [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+              {
+                v.add(-1.0 * gamma(j), G[i]);
+                ++j;
+              }
+            preconditioner.vmult(vhat, v);
+
+            uhat = vhat;
+            uhat *= omega;
+            j = 0;
+            for (unsigned int i = k; i < s; ++i)
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]:
```suggestion
            for (unsigned int i = k; i < s; ++i, ++j)
```

File: include/deal.II/lac/solver_idr.h [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+              {
+                uhat.add(gamma(j), U[i]);
+                ++j;
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: remove this line

File: include/deal.II/lac/solver_idr.h [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+              uhat.add(-1.0 * alpha, U[i]);
+            }
+          G[k] = ghat;
+          U[k] = uhat;
+
+          // Update kth column of M
+          for (unsigned int i = k; i < s; ++i)
+            M(i, k) = Q[i] * G[k];
+
+          // Orthoginalize r to Q0,...,Qk,
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]:
```suggestion
          // Orthogonalize r to Q0,...,Qk,
```

**Comment by tjhei** [comment]
great, thanks for cleaning it up and submitting this PR. Please fix the minor comments.

I would like to request somebody else to also review this PR as I am partial. ;-)

File: include/deal.II/lac/solver_idr.h (hunk 1 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: Following `solver_cg.h` and `solver_gmres.h`, I did not create a .cc file. Should I and move all implementations?

File: include/deal.II/lac/solver_idr.h (hunk 5 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: It is the number of vectors stored in the created object. I use 3 of these objects in the solve() function: U, G, and Q, each with `s` vectors. This class (like the temporary vectors in GMRES, I copied this class from solver_gmres.h) allows me to organize the vectors which depend on `s`.

File: include/deal.II/lac/solver_idr.h (hunk 6 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: This is actually left over from the GMRES implementation. I do not need this function and will delete it.

File: include/deal.II/lac/solver_idr.h (hunk 5 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: This is actually from the GMRES implementation and not needed here. I am removing it.

File: include/deal.II/lac/solver_idr.h [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+ * This class implements the IDR(s) method used for solving nonsymmetric,
+ * indefinite linear systems, developed in <a
+ * href="https://epubs.siam.org/doi/abs/10.1137/070685804">
+ * IDR(s): A Family of Simple and Fast Algorithms for Solving Large
+ * Nonsymmetric Systems of Linear Equations by Martin B. van Gijzen and Peter
+ * Sonneveld </a>. The implementation here is the preconditioned version from <a
+ * href="https://dl.acm.org/citation.cfm?id=2049667">
+ * Algorithm 913: An Elegant IDR(s) Variant that Efficiently Exploits
+ * Biorthogonality Properties
+ * by Martin B. van Gijzen and Peter Sonneveld</a>. The local structure @p AdditionalData takes
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]: Can you break this line manually?

File: include/deal.II/lac/solver_idr.h [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+ * the value for the parameter s which can be any integer greater than or equal
+ * to 1. For <code>s=1</code>, this method has similar convergence to BiCGStab.
+ *
+ * @note Each iteration of IDR(s) requires <code>s+1</code> preconditioning steps and matrix-vector
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]: Can you break this line manually?

File: include/deal.II/lac/solver_idr.h [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+    }
+
+
+
+    template <class VectorType>
+    inline VectorType &
+    TmpVectors<VectorType>::operator()(const unsigned int i,
+                                       const VectorType & temp)
+    {
+      Assert(i < data.size(), ExcIndexRange(i, 0, data.size()));
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]:
```suggestion
      AssertIndexRange(i,data.size());
```

File: include/deal.II/lac/solver_idr.h [hunk 14]
```diff
@@ -0,0 +1,475 @@
...
+
+
+
+template <class VectorType>
+SolverIDR<VectorType>::SolverIDR(SolverControl &           cn,
+                                 VectorMemory<VectorType> &mem,
+                                 const AdditionalData &    data)
+  : SolverBase<VectorType>(cn, mem)
+  , additional_data(data)
+
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]:
```suggestion
```

File: include/deal.II/lac/solver_idr.h [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+{}
+
+
+
+template <class VectorType>
+SolverIDR<VectorType>::SolverIDR(SolverControl &cn, const AdditionalData &data)
+  : SolverBase<VectorType>(cn)
+  , additional_data(data)
+
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]:
```suggestion
```

File: include/deal.II/lac/solver_idr.h [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+
+
+template <class VectorType>
+void
+SolverIDR<VectorType>::print_vectors(const unsigned int,
+                                     const VectorType &,
+                                     const VectorType &,
+                                     const VectorType &) const
+{}
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]: Any reason not to implement this? Won't `VectorType::print(std::cout);` work for all `VectorType`s?

File: include/deal.II/lac/solver_idr.h [hunk 17]
```diff
@@ -0,0 +1,475 @@
...
+
+      // Inner iteration over s
+      for (unsigned int k = 0; k < s; ++k)
+        {
+          // Solve M(k:s)*gamma = phi(k:s)
+          Vector<double> gamma(s - k);
+          {
+            Vector<double>            phik(s - k);
+            FullMatrix<double>        Mk(s - k, s - k);
+            std::vector<unsigned int> indices;
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]: All this looks a little bit suspicious, but is likely fine since `s` is small.

File: include/deal.II/lac/solver_idr.h [hunk 18]
```diff
@@ -0,0 +1,475 @@
...
+            unsigned int              j = 0;
+            for (unsigned int i = k; i < s; ++i, ++j)
+              {
+                indices.push_back(i);
+                phik(j) = phi(i);
+              }
+            Mk.extract_submatrix_from(M, indices, indices);
+
+            Mk.gauss_jordan();
```
Comment (masterleinad) [inline include/deal.II/lac/solver_idr.h]: ... and this more expensive anyway.

File: include/deal.II/lac/solver_idr.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/lac/solver_idr.h]: I guess we could consider doing that in another pull request (for all `Solver` classes).

File: include/deal.II/lac/solver_idr.h (hunk 16 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: I copied this function from `solver_cg.h`. In every `solver_*.h` file (except `solver_gmres.h` where there is no `print_vectors()` function) this function is implemented to do nothing. I never call this function so I will just remove it.

File: include/deal.II/lac/solver_idr.h (hunk 17 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: Yes, `s<10` is what is suggested in the paper I linked above (other papers similarly use single digit values).

File: include/deal.II/lac/solver_idr.h (hunk 16 above)
Comment (masterleinad) [reply include/deal.II/lac/solver_idr.h]: This function is there to be able to print intermediate solutions by using `DoFPrintSolverStep` with a solver class as base class. It might not be a bad idea to just call this function after updating the solution vector.

Separately, we also need tests for `DoFPrintSolverStep` anyway as there a none...

File: include/deal.II/lac/solver_idr.h (hunk 18 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: I will change to use `FullMatrix::invert` which hard codes up to 4x4 inverse, then uses `gauss_jordan()` if it is larger.

File: include/deal.II/lac/solver_idr.h (hunk 16 above)
Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: Ok, I'll keep it as is and add a `print_vectors()` call after each update like in `solver_cg.h`.

**Comment by tcclevenger** [comment]
Thanks, @masterleinad. I updated with the changes you suggested.

File: include/deal.II/lac/solver_idr.h [hunk 19]
```diff
@@ -0,0 +1,480 @@
...
+   */
+  SolverIDR(SolverControl &           cn,
+            VectorMemory<VectorType> &mem,
+            const AdditionalData &    data = AdditionalData());
+
+  /**
+   * Constructor. Use an object of type GrowingVectorMemory as a default to
+   * allocate memory.
+   */
+  SolverIDR(SolverControl &cn, const AdditionalData &data = AdditionalData());
```
Comment (tjhei) [inline include/deal.II/lac/solver_idr.h]: make this explicit

**Comment by tjhei** [comment]
Looks good to me as well.

File: include/deal.II/lac/solver_idr.h [hunk 20]
```diff
@@ -0,0 +1,481 @@
...
+  internal::SolverIDRImplementation::TmpVectors<VectorType> G(s, this->memory);
+  internal::SolverIDRImplementation::TmpVectors<VectorType> U(s, this->memory);
+  internal::SolverIDRImplementation::TmpVectors<VectorType> Q(s, this->memory);
+  FullMatrix<double>                                        M(s, s);
+
+  // Random number generator for vector entries of
+  // Q (normal distribution, mean=0 sigma=1)
+  boost::mt19937               rng;
+  boost::normal_distribution<> nd(0.0, 1.0);
+  boost::variate_generator<boost::mt19937 &, boost::normal_distribution<>>
```
Comment (bangerth) [inline include/deal.II/lac/solver_idr.h]: We now use the C++11 facilities here instead: https://en.cppreference.com/w/cpp/numeric/random
There are already other places in the library that use `std::mt19937` that you can grep for.

File: include/deal.II/lac/solver_idr.h [hunk 21]
```diff
@@ -0,0 +1,481 @@
...
+          }
+
+          // Update G and U
+          // Orthogonalize ghat to Q0,..,Q_{k-1}
+          // and update uhat
+          for (unsigned int i = 0; i < k; ++i)
+            {
+              double alpha = (Q[i] * ghat) / M(i, i);
+              ghat.add(-1.0 * alpha, G[i]);
+              uhat.add(-1.0 * alpha, U[i]);
```
Comment (bangerth) [inline include/deal.II/lac/solver_idr.h]: `-alpha` in both places :-)

**Comment by tcclevenger** [comment]
Thanks, @bangerth! Changes have been made.

File: include/deal.II/lac/solver_idr.h [hunk 22]
```diff
@@ -0,0 +1,479 @@
...
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/solver.h>
+#include <deal.II/lac/solver_control.h>
+
+#include <boost/random.hpp>
```
Comment (bangerth) [inline include/deal.II/lac/solver_idr.h]: Almost: You need to change the header file:
```suggestion
#include <random>
```

Comment (tcclevenger) [reply include/deal.II/lac/solver_idr.h]: Yes, of course!

**Comment by bangerth** [comment]
OK, now just squash and it's ready!

**Comment by tcclevenger** [comment]
> OK, now just squash and it's ready!

Done. Thanks, @bangerth!

**Comment by bangerth** [comment]
Oh, separately, could you please add an entry to `doc/news/changes/major`?

**Comment by bangerth** [comment]
Separate commit is ok.
